chalk-ir/src/lib.rs

```diff
@@ -979,6 +1013,60 @@ impl<I: Interner> ConcreteConst<I> {
     }
 }
 
+/// Unevaluated const expression, which can be evaluated
+#[derive(Copy, Clone, PartialEq, Eq, Hash, HasInterner)]
+pub struct UnevaluatedConst<I: Interner> {
+    /// The interned expression
+    pub interned: I::InternedUnevaluatedConst,
+}
+
+impl<I: Interner> UnevaluatedConst<I> {
+    /// Attempts to evaluate the const expression.
+    pub fn try_eval(&self, ty: &Ty<I>, interner: &I) -> Result<ConcreteConst<I>, ConstEvalError> {
+        Ok(ConcreteConst {
+            interned: interner.try_eval_const(&ty.interned, &self.interned)?,
+        })
+    }
+
+    /// Checks whether two unevaluated constants are equal.
+    pub fn const_eq(&self, ty: &Ty<I>, other: &UnevaluatedConst<I>, interner: &I) -> bool {
+        interner.unevaluated_const_eq(&ty.interned, &self.interned, &other.interned)
+    }
+}
+
+/// A mock unevaluated const expression.
+/// `Some(n)` evaluates to `n`,
+/// and `None` evaluates to a specific but unknown constant.
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub struct UnevaluatedConstData(pub Option<u32>);
+
+impl Debug for UnevaluatedConstData {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        match self.0 {
+            Some(n) => write!(fmt, "{}?", n)?,
+            None => write!(fmt, "?")?,
+        }
+        Ok(())
+    }
+}
+
+/// An error that can occur when evaluating a const expression.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum ConstEvalError {
```

Comment:
Probably also want an `Unevaluatable`. Though I don't exactly know how Chalk would handle these differently. (`TooGeneric` is obvious `Ambiguous`, but would `Unevaluable` return `Ambiguous` or `NoSolution`?)